website/docs/guides/serverless-with-aws-lambda-and-api-gateway.html.md

```diff
@@ -0,0 +1,634 @@
+with Terraform. Since our configuration requires a version number as an
+input variable, provide a placeholder value to destroy:
+
+```
+$ terraform destroy -var="app_version=0.0.0"
+```
+
+Since the artifact zip files and the S3 bucket itself were created
+outside of Terraform, they must also be cleaned up outside of Terraform. This
+can be done via [the S3 console](https://s3.console.aws.amazon.com/s3/home).
+Note that all of the objects in the bucket must be deleted before the bucket
+itself can be deleted.
+
+## Further Reading
+
+The following Terraform resource types are used in this tutorial:
+
+* [`aws_lambda_function`](/docs/providers/aws/r/aws_lambda_function.html)
+* [`aws_lambda_permission`](/docs/providers/aws/r/aws_lambda_permission.html)
+* [`aws_api_gateway_rest_api`](/docs/providers/aws/r/aws_api_gateway_rest_api.html)
+* [`aws_api_gateway_resource`](/docs/providers/aws/r/aws_api_gateway_resource.html)
+* [`aws_api_gateway_method`](/docs/providers/aws/r/aws_api_gateway_method.html)
+* [`aws_api_gateway_integration`](/docs/providers/aws/r/aws_api_gateway_integration.html)
+* [`aws_iam_role`](/docs/providers/aws/r/aws_iam_role.html)
+
+The reference page for each resource type provides full details on all of its
+supported arguments and exported attributes.
+
+### Custom Domain Names and TLS Certificates
+
+For the sake of example, this guide uses the test URLs offered by default
+by API Gateway. In practice, most applications will be deployed at a custom
+hostname.
+
+To use a custom domain name you must first register that domain and configure
+DNS hosting for it. You must also either create an
+[Amazon Certificate Manager](https://aws.amazon.com/certificate-manager/)
+certificate or register a TLS certificate with a third-party certificate
+authority.
+
+Configuring the domain name is beyond the scope of this tutorial, but if
+you already have a hostname and TLS certificate you wish to use then you can
+register it with API Gateway using the
+[`aws_api_gateway_domain_name`](/docs/providers/aws/r/api_gateway_domain_name.html)
+resource type.
+
+A registered domain name is then mapped to a particular "REST API" object using
+[`aws_api_gateway_base_path_mapping`](/docs/providers/aws/r/api_gateway_base_path_mapping.html).
+The configured domain name then becomes an alias for a particular deployment
+stage.
+
+### Making Changes to the API Gateway Configuration
+
+This guide creates a very simple API Gateway Configuration with a single
+resource that passes through all requests to a single destination. The upgrade
+steps then modify only the AWS Lambda function, leaving the API Gateway
+configuration unchanged.
+
+Due to API Gateway's staged deployment model, if you _do_ need to make changes
+to the API Gateway configuration you must explicitly request that it be
+re-deployed by "tainting" the deployment resource:
+
+```
+$ terraform taint aws_api_gateway_deployment.example
+```
```

Comment:
Can we also mention that such redeployment will incur downtime here so folks are not surprised? I don't know what's the best wording, but this is mainly due to complex API and Terraform implementation which will (hopefully) change in the future so we can do more sane re-deployments of new versions with zero downtime.

Reply:
The downtime here is because the stage gets temporarily "un-deployed"? I didn't actually know that :grinning: so I guess I was just lucky that the downtime I've caused doing this occasionally in the past was not long enough for anyone to notice... or maybe having a CDN in front saved me.

In any case, yes... I will add a note about that here. I went to look at [the code](https://github.com/terraform-providers/terraform-provider-aws/blob/cefb0b0ce7f304853e6f3cb66880dbae1cc4cb2c/aws/resource_aws_api_gateway_deployment.go#L175-L188) to see what it's doing, and I think I'll just document plainly what's going to happen if you do this and thus leave it up to the reader to decide if they find that acceptable.